### ISSUE TITLE
(auto)Bounds related warning that can be improved (2.12 and earlier)

### ISSUE THREAD

**Issue by jypeter** [UV-CDAT/cdms#189]
@dnadeau4 I got the following warning, that can probably be improved
```
>>> v1 =  f('LEAF_M_c')
/home/share/unix_files/cdat/miniconda2/envs/uvcdat-2.12/lib/python2.7/site-packages/cdms2/axis.py:1617: UserWarning:
Your first bounds[0,0] -180.000000000000000 will be corrected to -180.000000000000000
Your bounds bounds[-1,1] 180.000000000000000 will be corrected to 180.000000000000000
  warnings.warn(msg, UserWarning)
```
- **-180** corrected to **-180** (and same thing for 180) is not very clear for the user! Especially since there is no error
- Should `Your bounds bounds[-1,1]` be "your **last** bounds[-1,1]" ?

Some information about the data:
- Definition of the variable
```
        float LEAF_M_n(time_counter, veget, lat, lon) ;
                LEAF_M_n:long_name = "Leaf mass" ;
                LEAF_M_n:units = "gN/m^2 " ;
                LEAF_M_n:online_operation = "average" ;
                LEAF_M_n:interval_operation = "1800 s" ;
                LEAF_M_n:interval_write = "1 yr" ;
                LEAF_M_n:cell_methods = "time: mean (interval: 1800 s)" ;
                LEAF_M_n:_FillValue = 9.96921e+36f ;
                LEAF_M_n:missing_value = 9.96921e+36f ;
                LEAF_M_n:coordinates = "time_centered veget" ;

```
- **lon** axis has _no defined bounds in the file_, and has the following values
```
 lon = -179, -177, -175, -173, -171, -169, -167, -165, -163, -161, -159,
    -157, -155, -153, -151, -149, -147, -145, -143, -141, -139, -137, -135,
    -133, -131, -129, -127, -125, -123, -121, -119, -117, -115, -113, -111,
    -109, -107, -105, -103, -101, -99, -97, -95, -93, -91, -89, -87, -85,
    -83, -81, -79, -77, -75, -73, -71, -69, -67, -65, -63, -61, -59, -57,
    -55, -53, -51, -49, -47, -45, -43, -41, -39, -37, -35, -33, -31, -29,
    -27, -25, -23, -21, -19, -17, -15, -13, -11, -9, -7, -5, -3, -1, 1, 3, 5,
    7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29, 31, 33, 35, 37, 39, 41, 43,
    45, 47, 49, 51, 53, 55, 57, 59, 61, 63, 65, 67, 69, 71, 73, 75, 77, 79,
    81, 83, 85, 87, 89, 91, 93, 95, 97, 99, 101, 103, 105, 107, 109, 111,
    113, 115, 117, 119, 121, 123, 125, 127, 129, 131, 133, 135, 137, 139,
    141, 143, 145, 147, 149, 151, 153, 155, 157, 159, 161, 163, 165, 167,
    169, 171, 173, 175, 177, 179 ;
```

Note that I don't get a warning if I use _cdms2.setAutoBounds('off')_, and that the generated bounds (when not using 'off') are correct:
```
>>> v1.getLongitude().getBounds()
array([[-180., -178.],
       [-178., -176.],
       [-176., -174.],
...
       [ 174.,  176.],
       [ 176.,  178.],
       [ 178.,  180.]])
```

**Comment by jypeter** [comment]
I forgot to mention that the lon axis seems to be defined correctly
```
        float lon(lon) ;
                lon:axis = "X" ;
                lon:standard_name = "longitude" ;
                lon:long_name = "Longitude" ;
                lon:units = "degrees_east" ;
```
I opened this file with cdms2 because it breaks cdo **randomly** (on different servers!). But cdms2 did not complain. It's an output of the ORCHIDEE vegetation model, but they are kind of used to having weird things... I can temporarily upload the 102M sample file somewhere if @dnadeau4 wants a file that breaks cdo to play with

**Comment by dnadeau4** [comment]
@jypeter I think I fixed that with Karl Taylor a couple  of months ago.

**Comment by dnadeau4** [comment]
https://github.com/UV-CDAT/cdms/issues/126

**Comment by jypeter** [comment]
@dnadeau4 the message looks very similar indeed. Is it normal that I'm still getting this in 2.12 ?

I don't know if this is a bug or a feature. cdms2 should not say that it is changing something, when it is not changing anything! Or maybe it should say more explicitly why it is doing it ?

In some rare cases, I also read or plot data using `v = f('varname', longitude=(-180, 360, 'co')`, so cdms2 needs to be able to keep on working even when 'lon[-1] - lon[0] >= 360"

**Comment by doutriaux1** [comment]
@jypeter I agree cdms should NOT change your data, I believe this is just a round up error. @dnadeau4 why did you implement this for originally? I can't remember exactly why this became an issue?

**Comment by dnadeau4** [comment]
I implemented this due to CMOR3 converting from -180,180 to 0-360.  We had monotonic  issues during the conversion due to floating point precision,  we snap the bounds to the closes line. 

@taylor13 can comment on this.
https://github.com/PCMDI/cmor/issues/146

**Comment by doutriaux1** [comment]
@dnadeau4 should we turn it off by default? And just give the user a function to do it? I usually don't like changing the user's data.

**Comment by dnadeau4** [comment]
We are not really changing the data just adjust the precision, from `float32` to `double`.    @taylor13 and I worked hard on this lat spring. @JimBiardCics  needs this feature.   This is only when `autobound`s is "on" and it is "off" by default.  

I always disliked `autobounds`, it is prone to all kind of errror.

**Comment by taylor13** [comment]
As I recall, I suggested that when autobounds is on and the domain spans 360 degrees longitude  (with some allowance for a small error), then CDMS should adjust the bounds to *exactly* span 360 degrees.  Furthermore, if the bounds of the domain spanned 360 degrees and one bound was very close to an integer, then the domain bounds should be adjusted to coincide with the nearest meridian.  For example if the domain bounds were -179.999 and 180.000, the algorithm should adjust these to -180.000 and 180.000.  Similarly, -179.999, 180.001 would also be adjusted to -180, 180, and 0.0, 359.999 would be adjusted to 0, 360, and -90.001, 270.001 should be adjusted to -90, 270.

I thought that is what Denis implemented, but if I understand #126 which says:
```
When values of longitude are 0-360, periodicity is warning users about changing bounds, 
bounds should only be adjusted when values are from -180,180.
```
maybe my suggestion was not implemented (or only implemented for the special case of the domain approximately spanning -180 to 180).  I don't understand why the case -180 to 180 is special.  I think that when autobounds is on, the adjustment to span 360 should always be done (when the domain approximately spans 360).

**Comment by doutriaux1** [comment]
@dnadeau4 I believe autobounds is on by default for lat/lon though. But anyway your explanation makes sense. I'm cool with leaving it as is.

**Comment by dnadeau4** [comment]
@jypeter Do you have lon_bnds in your file?

**Comment by taylor13** [comment]
According to https://github.com/UV-CDAT/cdms/issues/189#issue-271896070 . the answer is "no".
